src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,1077 @@
+            type.SomeProp1 &&
+            !type.SomeProp2 &&
+            !type.SomeProp3 && 
+            !(type.SomeProp4 &&
+              type.SomeProp5 &&
+              type.SomeProp6) &&
+            (type.Kind == 4 || type.Kind == 5 || type.Kind == 6) &&
+            type.ContainingSymbol is var declContainer &&
+            declContainer != null &&
+            declContainer is C &&
+            type.ContainingSymbol.ContainingSymbol is var containingSymbol &&
+            ((C)declContainer).Kind is int &&
+            ((C)declContainer).Kind == 0 &&
+            ((C)declContainer).Name == 123 [||]&& a; 
+    }
+    int Name;
+    C ContainingSymbol;
+    int Kind;
+    int? Nullable1;
+    int? Nullable2;
+    int? Nullable3;
+    int? Nullable4;
+    bool SomeProp1;
+    bool SomeProp2;
+    bool SomeProp3;
+    bool SomeProp4;
+    bool SomeProp5;
+    bool SomeProp6;
+}",
+@"class C
+{
+    bool M(C type, bool a)
+    {
+        return
+            a && type is
+        {
+            SomeProp1: true,
+            SomeProp2: false,
+            SomeProp3: false,
+            SomeProp4: false,
+            SomeProp5: false,
+            SomeProp6: false,
+            Name: >= 1 and <= 3,
+            Nullable1: not null,
+            Nullable2: null,
+            Nullable3: 5,
+            Kind: >= 4 and <= 6,
+            ContainingSymbol: C
+            {
+                ContainingSymbol: var containingSymbol,
+                Kind: int and 0,
+                Name: 123,
+            } declContainer,
+        }; 
+    }
+    int Name;
+    C ContainingSymbol;
+    int Kind;
+    int? Nullable1;
+    int? Nullable2;
+    int? Nullable3;
+    int? Nullable4;
+    bool SomeProp1;
+    bool SomeProp2;
+    bool SomeProp3;
+    bool SomeProp4;
+    bool SomeProp5;
+    bool SomeProp6;
+}");
+        }
+
+        [Fact]
+        public async Task Test04()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C type)
+    {
+        return type.Name == ""ValueTuple"" &&
+            type.ContainingSymbol is SomeType declContainer &&
+            declContainer.Kind == 0 [||]&&
+            declContainer.Name == ""System"";
+    }
+    string Name;
+    C ContainingSymbol;
+    int Kind;
+    class SomeType : C {}
+}",
+@"class C
+{
+    bool M(C type)
+    {
+        return type is
+        {
+            Name: ""ValueTuple"",
+            ContainingSymbol: SomeType { Kind: 0, Name: ""System"" } declContainer,
+        };
+    }
+    string Name;
+    C ContainingSymbol;
+    int Kind;
+    class SomeType : C {}
+}");
+        }
+
+        [Fact]
+        public async Task Test05()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M(C type)
+    {
+        return type.Name == ""ValueTuple"" [||]&& type.IsStruct;
+    }
+    string Name;
+    bool IsStruct;
+}",
+@"class C
+{
+    bool M(C type)
+    {
+        return type is { IsStruct: true, Name: ""ValueTuple"" };
```

Comment:
seems unfortunate that the order of the checks is changed.  intentional?  desirable?

Reply:
Since we're piping all this through a PooledDictionary the order can change. A possible solution is to sort by original SpanStart, maybe?

Reply:
yup.  sounds liek a great idea.  let's do that :)

Reply:
and, sinec you mentioned its' a dictionary, this is now doubly important for me that we do this.  I do not want behavior to change as the dictionary impl changes in the future.  we should be isolated for non-deterministic behavior.